Composer/packages/client/src/pages/botProject/skill-configuration/AllowedCallers.tsx

```diff
@@ -0,0 +1,170 @@
+  const itemRef = React.useRef<ITextField | null>(null);
+  const didMount = React.useRef<boolean>(false);
+
+  React.useEffect(() => {
+    if (!value && !didMount.current) {
+      itemRef.current?.focus();
+    }
+    didMount.current = true;
+  }, []);
+
+  return (
+    <Row horizontal>
+      <Input
+        componentRef={(ref) => (itemRef.current = ref)}
+        styles={textFieldStyles}
+        value={value}
+        onBlur={onBlur}
+        onChange={onChange}
+      />
+      <IconButton aria-label={formatMessage('Remove item')} iconProps={{ iconName: 'Trash' }} onClick={onRemove} />
+    </Row>
+  );
+});
+
+type Props = {
+  projectId: string;
+};
+
+export const AllowedCallers: React.FC<Props> = ({ projectId }) => {
+  const { setSettings } = useRecoilValue(dispatcherState);
+  const rootBotProjectId = useRecoilValue(rootBotProjectIdSelector) || '';
+  const settings = useRecoilValue(settingsState(projectId));
+  const mergedSettings = mergePropertiesManagedByRootBot(projectId, rootBotProjectId, settings);
+  const { skillConfiguration } = mergedSettings;
+
+  const setAllowedCallers = React.useCallback(
+    (allowedCallers: string[] = []) => {
+      const updatedSetting = {
+        ...cloneDeep(mergedSettings),
+        skillConfiguration: { ...skillConfiguration, allowedCallers },
+      };
+      setSettings(projectId, updatedSetting);
+    },
+    [mergedSettings, projectId, skillConfiguration]
+  );
+
+  const onBlur = React.useCallback(() => {
+    const updatedAllowedCallers = [...skillConfiguration?.allowedCallers];
+    setAllowedCallers(updatedAllowedCallers.filter(Boolean));
+  }, [skillConfiguration?.allowedCallers, setAllowedCallers]);
+
+  const onChange = React.useCallback(
+    (index: number) => (_: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>, newValue = '') => {
+      const updatedAllowedCallers = [...skillConfiguration?.allowedCallers];
+      updatedAllowedCallers[index] = newValue;
+      setAllowedCallers(updatedAllowedCallers);
+    },
+    [skillConfiguration?.allowedCallers, setAllowedCallers]
+  );
+
+  const onRemove = React.useCallback(
+    (index: number) => () => {
+      const updatedAllowedCallers = skillConfiguration?.allowedCallers?.filter((_, itemIndex) => itemIndex !== index);
+      setAllowedCallers(updatedAllowedCallers);
+    },
+    [skillConfiguration?.allowedCallers, setAllowedCallers]
+  );
+
+  const onAddNewAllowedCaller = React.useCallback(() => {
+    setAllowedCallers([...skillConfiguration?.allowedCallers, '']);
+  }, [skillConfiguration?.allowedCallers, setAllowedCallers]);
+
+  return (
+    <React.Fragment>
+      <div css={tableColumnHeader()}>{formatMessage('Allowed callers')} </div>
+      <ItemContainer>
+        {skillConfiguration?.allowedCallers?.map((caller, index) => {
+          return (
+            <Item key={index} value={caller} onBlur={onBlur} onChange={onChange(index)} onRemove={onRemove(index)} />
```

Comment:
nit: better not to use index as key

Reply:
Suggestions?

Reply:
caller

Reply:
Caller is not necessarily unique, and it’s not constant.